packages/winpanda/extra/src/winpanda/common/cli.py

```diff
@@ -0,0 +1,73 @@
+"""Panda package management for Windows.
+
+Application CLI specification.
+"""
+
+
+class CLI_COMMAND:
+    """CLI command set."""
+    SETUP = 'setup'
+    TEARDOWN = 'teardown'
+    START = 'start'
+    STOP = 'stop'
```

Comment:
How does this set of commands relate to the `pkgpanda` commands? https://github.com/dcos/dcos/blob/f4899e6d2dbf2a3a3a5d94c5da347796275852ef/pkgpanda/cli.py#L4-L13

Reply:
Here there is a number of points to take into account:

1) "winpanda setup" - this directly corresponds "pkgpanda setup" functionality.

2) "winpanda teardown" -  this one directly corresponds to the "pkgpanda uninstall", which does completely remove entire DC/OS installation (stops all the system services and removes all the appropriate configs for systemd, removes all the files/dirs related to a DC/OS installation). 
    I just think that the "teardown" is a better fit for the operation name here.  It can be considered as a synonym  for the "uninstall" but we use "install/uninstall" names also at the package and the system service levels and I just wanted to use another word for the whole DC/OS operation level for  better separation. 
    The other point is that we use the "setup", but not the "install" name for  the whole DC/OS deployment operation and the "teardown" better fits as the name for opposite operation. 
    I can rename it to "uninstall", if the considerations above do not make sense for you.

3) "winpanda start/stop" - this will do start/stop all the system service processes corresponding to all the installed DC/OS packages. A kind of DC/OS "on/off" for a windows-based node. This is different to pkgpanda behaviour. 
    The pkgpanda does these operations under the hood for top level DC/OS installation management operations like "setup" and "activate", but does not expose them via CLI. 
    I think that this will be useful to implement a "maintenance" state for windows nodes when we need to repair/investigate issues, and also will facilitate the DC/OS "upgrade" operation.
   
4) As far as I can see, currently only three of the whole bunch of operations which the pkgpanda exposes via CLI are really used. These are "setup", "fetch" and "activate". The former one is used for the clean DC/OS installation on a node in the dcos_install.sh script (/opt/dcos-install/<DC/OS version>/genconf/serve/dcos_install.sh residing on a bootsrap node), two latter are used during DC/OS upgrade operation in the dcos_node_upgrade.sh script (/opt/dcos-install/<DC/OS version>/genconf/serve/upgrade_from_<DC/OS version>/<UUID>/dcos_node_upgrade.sh residing on a bootstrap node).
    I can assume that the others were intended to be used via pkgpanda web-API,  but in accordance with our initial investigation of pkgpanda operation on a node (Artem said that, as far as I can remember) that API is not in use now and unlikely will be used in the future.
Ref: https://docs.d2iq.com/mesosphere/dcos/1.13/administering-clusters/component-management/#routes

    Hence the question: do we need to replicate the complete pkgpanda's  node-level functionality in winpanda?
   
5) We had an internal discussions within the local team and preliminary agreed on the logic of DC/OS upgrade operation. Our vision is that it will be  covered by the "setup" operation as well. 
    If we introduce a notion of DC/OS installation state (i.e. list of packages that are currently active on a node), than both  "upgrade" and **clean** "setup" operations in their essence are just two particular cases of the **generic** "setup" operation. 
    Meaning that the  **clean** "setup" is the case when the state is empty (list of active packages is empty) and so it needs to install all the packages listed as parts of a DC/OS version being deployed (e.g. specified in a special <DC/OS version>--package_list.json file at a  bootstrap node).
    The "upgrade" operation is the case when we have the DC/OS state non-empty (list of active packages is non empty) and so we need to compare it with a list describing a new DC/OS version (the one we want to upgrade to, e.g <New DC/OS version>--package_list.json at a bootstrap node), produce a list of packages to be removed and another list of those to be installed, and then just do remove/install appropriately for the every package in those lists.
    This approach would save us from generating additional helper scripts like the dcos_node_upgrade.sh, avoiding unnecessary complication and sharing responsibility of DC/OS upgrade operation between the "upgrade script" and winpanda app, and alow us to have just a single artifact, describing a set of  packages that are to be present on a node in accordance with desired DC/OS version - <DC/OS version>--package_list.json. 
    Also moving all the responsibility of the "upgrade" operation to the winpanda app would allow us to provide more consistent error reporting and logging for that operation (from the single source).
   
    As far as I can understand now a bootstrap node already provides an equivalent of that imaginary "<DC/OS version>--package_list.json" artifact as the /opt/dcos-install/<DC/OS version>/genconf/serve/package_lists/UUID.package_list.json. 
    So, the idea outlined above is no way a completely new approach to define the   **desired** DC/OS state. And requires very little, or no changes to the logic iplemented at the bootstrap node side. We just implement logic of winpanda app to handle the "upgrade" operation more wisely that the pkgpanda does.